### PR TITLE
DSPDC-1155 Port to enumeratum for generated categorical data.

### PR DESCRIPTION
**Pull request by danxmoran**
Still need to add back Scaladoc, but this is the general idea.

### CODE REVIEW

File: hle-survey/transformation/src/main/scala/org/broadinstitute/monster/dap/DogTransformations.scala [hunk 1]
```diff
@@ -6,20 +6,22 @@ import org.broadinstitute.monster.dogaging.jadeschema.table.HlesDog
...
-      Some(ResidentialEnvironmentTransformations.mapResidentialEnvironment(rawRecord)),
-    hlesDogBehavior = Some(BehaviorTransformations.mapBehavior(rawRecord)),
-    hlesDogDiet = Some(DietTransformations.mapDiet(rawRecord)),
-    hlesDogMedsPreventatives =
-      Some(MedsAndPreventativesTransformations.mapMedsPreventatives(rawRecord)),
-    hlesDogHealthSummary = Some(HealthStatusTransformations.mapHealthSummary(rawRecord)),
-    hlesDogFutureStudies = Some(AdditionalStudiesTransformations.mapFutureStudies(rawRecord))
-  )
+  def mapDog(rawRecord: RawRecord): HlesDog =
+    HlesDog(
```
Comment (danxmoran) [inline hle-survey/transformation/src/main/scala/org/broadinstitute/monster/dap/DogTransformations.scala:10]: Unrelated formatting from bumping scalafmt

File: hle-survey/transformation/src/main/scala/org/broadinstitute/monster/dap/HealthTransformations.scala [hunk 2]
```diff
@@ -1,301 +1,94 @@
...
 
+import org.broadinstitute.monster.dap.healthcondition.HealthCondition
 import org.broadinstitute.monster.dogaging.jadeschema.table.HlesHealthCondition
 
 object HealthTransformations {
 
   /** Parse all health-condition-related fields out of a raw RedCap record. */
   def mapHealthConditions(rawRecord: RawRecord): Iterable[HlesHealthCondition] =
-    healthConditions.flatMap(processHealthCondition(rawRecord, _))
+    HealthCondition.values.flatMap { healthCondition =>
```
Comment (danxmoran) [inline hle-survey/transformation/src/main/scala/org/broadinstitute/monster/dap/HealthTransformations.scala:10]: This logic is rewritten to use the new structure of the enum values.

File: hle-survey/transformation/src/main/scala/org/broadinstitute/monster/dap/healthcondition/HealthCondition.scala [hunk 3]
```diff
@@ -0,0 +1,105 @@
+package org.broadinstitute.monster.dap.healthcondition
+
+import enumeratum.values.{LongEnum, LongEnumEntry}
+
+sealed abstract class HealthCondition(
+  override val value: Long,
+  val label: String,
+  val conditionType: HealthConditionType,
+  val abbreviation: String,
```
Comment (danxmoran) [inline hle-survey/transformation/src/main/scala/org/broadinstitute/monster/dap/healthcondition/HealthCondition.scala:9]: I am making an assumption here that the "core" abbreviation for a disease is consistent across cg/dx cases, even if the prefix/suffix vary. It's true for all the cases we'd already coded / for other cases I spot-checked.

File: hle-survey/transformation/src/main/scala/org/broadinstitute/monster/dap/healthcondition/HealthCondition.scala [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+  val hasCg: Boolean,
+  val hasDx: Boolean,
+  val isOther: Boolean = false,
+  val cgPrefixOverride: Option[String] = None,
+  val computeGate: String => String = identity
```
Comment (danxmoran) [inline hle-survey/transformation/src/main/scala/org/broadinstitute/monster/dap/healthcondition/HealthCondition.scala:14]: These bottom two args are escape hatches for dealing with annoying cases. I'm not super happy with using them, but I figured it wasn't worth battling the terrible RedCap layout any longer.

File: hle-survey/transformation/src/main/scala/org/broadinstitute/monster/dap/healthcondition/HealthConditionType.scala [hunk 5]
```diff
@@ -0,0 +1,44 @@
+package org.broadinstitute.monster.dap.healthcondition
+
+import enumeratum.values.{LongEnum, LongEnumEntry}
+
+sealed abstract class HealthConditionType(
+  override val value: Long,
+  val label: String,
+  val cgKey: Option[HealthConditionType.CgKey],
+  val dxKey: Option[HealthConditionType.DxKey]
```
Comment (danxmoran) [inline hle-survey/transformation/src/main/scala/org/broadinstitute/monster/dap/healthcondition/HealthConditionType.scala:9]: This information was previously set on a per-condition basis, but it was largely copy-pasted since it's really a property of each "block" of conditions.

File: hle-survey/transformation/src/main/scala/org/broadinstitute/monster/dap/healthcondition/HealthConditionType.scala [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  case object Reproductive extends HealthConditionType(10L, "Reproductive", Some(CgKey("repro")), Some(DxKey("repro")))
+  case object Orthopedic extends HealthConditionType(11L, "Bone/Orthopedic", Some(CgKey("bones")), Some(DxKey("ortho")))
+  case object Neurologic extends HealthConditionType(12L, "Brain/Neurologic", Some(CgKey("brain")), Some(DxKey("neuro")))
+  case object Endocrine extends HealthConditionType(13L, "Endocrine", Some(CgKey("endocr")), Some(DxKey("endo")))
+  case object Hematopoietic extends HealthConditionType(14L, "Hematopoietic", Some(CgKey("blood")), Some(DxKey("hema")))
+  case object OtherCongenital extends HealthConditionType(15L, "Other congenital disorder", Some(CgKey("other", false)), None)
+  case object Infection extends HealthConditionType(16L, "Infectious disease", None, Some(DxKey("infectious", false)))
+  case object ToxinConsumption extends HealthConditionType(17L, "Toxin Consumption", None, Some(DxKey("dx_tox")))
+  case object Trauma extends HealthConditionType(18L, "Trauma", None, Some(DxKey("trauma")))
+  case object Immune extends HealthConditionType(19L, "Immune", None, Some(DxKey("immune")))
```
Comment (danxmoran) [inline hle-survey/transformation/src/main/scala/org/broadinstitute/monster/dap/healthcondition/HealthConditionType.scala:43]: Went ahead and filled this out all the way, to avoid future confusion about ordering.

File: hle-survey/transformation/src/main/scala/org/broadinstitute/monster/dap/healthcondition/HealthCondition.scala [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+) extends LongEnumEntry
+
+object HealthCondition extends LongEnum[HealthCondition] {
+  override val values = findValues
+
+  import HealthConditionType.{findValues => _, _}
+
+  // scalafmt: { maxColumn = 140, newlines.topLevelStatements = [] }
```
Comment (snwessel) [inline hle-survey/transformation/src/main/scala/org/broadinstitute/monster/dap/healthcondition/HealthCondition.scala:22]: Wow, this is just changing the formatting for this section?... Pretty neat! I didn't know that was possible.

File: hle-survey/transformation/src/main/scala/org/broadinstitute/monster/dap/healthcondition/HealthCondition.scala [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+  case object RMSF extends HealthCondition(1633L, "Rocky Mountain Spotted Fever (RMSF)", Infection, "rmsf", false, true)
+  case object Roundworms extends HealthCondition(1634L, "Roundworms", Infection, "rndworm", false, true)
+  case object Salmonellosis extends HealthCondition(1635L, "Salmonellosis", Infection, "slmosis", false, true)
+  case object SalmonPoison extends HealthCondition(1636L, "Salmon poisoning", Infection, "slmpois", false, true)
+  case object Tapeworms extends HealthCondition(1637L, "Tapeworms", Infection, "tpworm", false, true)
+  case object Toxoplasma extends HealthCondition(1638L, "Toxoplasma", Infection, "toxop", false, true)
+  case object Tularemia extends HealthCondition(1639L, "Tularemia", Infection, "tular", false, true)
+  case object Whipworms extends HealthCondition(1640L, "Whipworms", Infection, "whpworm", false, true)
+  case object OtherInfection
+      extends HealthCondition(1698L, "Other infectious disease", Infection, "infect_other", false, true, isOther = true)
```
Comment (snwessel) [inline hle-survey/transformation/src/main/scala/org/broadinstitute/monster/dap/healthcondition/HealthCondition.scala:143]: Also, I like this numbering system 👍
